addons/sale_purchase/models/sale_order_line.py

```diff
@@ -285,10 +286,8 @@ def _purchase_service_generation(self):
             will not create a second one.
         """
         sale_line_purchase_map = {}
-        for line in self:
-            line = line.with_company(line._purchase_service_get_company())
-            # Do not regenerate PO line if the SO line has already created one in the past (SO cancel/reconfirmation case)
-            if line.product_id.service_to_purchase and not line.purchase_line_count:
-                result = line._purchase_service_create()
-                sale_line_purchase_map.update(result)
+        lines_to_process = self.with_company(self._purchase_service_get_company()).filtered(
+            lambda l: l.product_id.service_to_purchase and not l.purchase_line_count)
+        if lines_to_process:
+            sale_line_purchase_map.update(lines_to_process._purchase_service_create())
```

Comment:
I am not sure to understand why you change that code. It seems your changes are equivalent to the previous code except you did it in batch, right? 🤔 

If I am right then why is it needed?

Reply:
> I am not sure to understand why you change that code. It seems your changes are equivalent to the previous code except you did it in batch, right? 🤔
> 
> If I am right then why is it needed?

Our purpose is to improve performance. Currently,` _purchase_service_create` is triggered multiple times. Instead, it will now be triggered only once.

Reply:
It is not a bad idea but we have to make sure the behavior will not be altered.

For instance, before your changes, we call `_purchase_service_get_company` method line per line, I guess it is to make sure we will have only one company because in some cases we could maybe have more than one company?
Here with your changes you assume, we will have only one company in all purchase lines contained in `self`. Hopefully, for the moment, that method is only called for each SO confirmed but if one day, it will not longer be the case, then you work could crash since `with_company` just expects to have one company.

Hello @clesgow 🙂 

what do you think? 🙂

Reply:
> Agree with you @xavierbol. I guess you could run into the issue when confirming multiple sale orders from different companies at once. For safety's sake, if you want to batch these then I'd recommend grouping them by `company_id`.

I have reverted this change. Currently, multiple sale orders cannot be confirmed from the list view. However, there might be other ways, such as using a cron job or alternative methods.